modules/@apostrophecms/rich-text-widget/index.js

```diff
@@ -205,7 +212,38 @@ module.exports = {
         return widget.content;
       },
 
+      // Handle permalinks
       async load(req, widgets) {
+        const widgetsByDocId = new Map();
+        let ids = [];
+        const project = self.options.project;
+        for (const widget of widgets) {
+          if (!widget.permalinkIds) {
+            return;
+          }
+          for (const id of widget.permalinkIds) {
+            const docWidgets = widgetsByDocId.get(id) || [];
+            docWidgets.push(widget);
+            widgetsByDocId.set(id, docWidgets);
+            ids.push(id);
+          }
+        }
+        ids = [ ...new Set(ids) ];
+        if (!ids.length) {
+          return;
+        }
+        const docs = await self.apos.doc.find(req, {
+          aposDocId: {
+            $in: ids
+          }
+        }).project(project).toArray();
+        for (const doc of docs) {
+          const widgets = widgetsByDocId.get(doc.aposDocId) || [];
+          for (const widget of widgets) {
+            widget._permalinkDocs = widget._permalinkDocs || [];
+            widget._permalinkDocs.push(doc);
+          }
+        }
```

Comment:
this whole function could be re-written with maps for a better readability, but I assume this is copied from A2, and it already works as it is